### PR TITLE
Locator discovery refactoring

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR refactors the LocatorDiscovery plugin in C#.

Main changes:
- reimplemented the Locator servant to forward request with a combination of ForwardAsync (when the encoding of the incoming request is the same as the discovered locator proxy's encoding) and "manual" forwarding otherwise - in this case the servant unmarshals the request and remarshals it with the discovered locator proxy's encoding.

- moved plugin from C# namespace ZeroC.IceLocatorDiscovery to ZeroC.Ice.LocatorDiscovery

- moved plugin Slice definitions from Slice module IceLocatorDiscovery to Slice module Ice::LocatorDiscovery

In general, it's not possible move Slice definitions from one module to another (or rename Slice interfaces) without breaking interop with existing versions. However it's not a concern for LocatorDiscovery since all calls to the LocatorDiscovery interfaces use UDP. Since it's all oneway, there is no calls to ice_isA, ice_id, ice_ids, and changing the type IDs of the Slice interfaces doesn't break anything.

### CODE REVIEW

**Comment by pepone** [comment]
Isn't the forwarding triggering a second locator lookup when `GetLocatorAsync` is called from `ForwardRequestAsync`

For example, if `DispatchAsync` gets a null locator it calls the base DispatchAsync, and then `ForwardRequestAsync` and the call to `GetLocatorAsync` will have to start looking up the locator because it is null.

**Comment by bernardnormier** [comment]
> Isn't the forwarding triggering a second locator lookup when `GetLocatorAsync` is called from `ForwardRequestAsync`
> 
> For example, if `DispatchAsync` gets a null locator it calls the base DispatchAsync, and then `ForwardRequestAsync` and the call to `GetLocatorAsync` will have to start looking up the locator because it is null.

Yes, that's correct. However, `GetLocatorAsync` caches the discovered locator (when not null), so the double discovery occurs only when the first discovery results in a null locator. 

Is this really a concern? Basically it means the failure scenario (could not discover any locator) takes longer to fail since LocatorDiscovery performs full 2 discoveries. If it's a concern, maybe I can rework the code to avoid it.

**Comment by pepone** [comment]
>Is this really a concern? Basically it means the failure scenario (could not discover any locator) takes longer to fail since LocatorDiscovery performs 2 full discoveries. If it's a concern, maybe I can rework the code to avoid it.


I think it is fine as is, the setting of `_nextRetry` should prevent the second try already

**Review by pepone — APPROVED** [review]

File: slice/IceLocatorDiscovery/IceLocatorDiscovery.ice [hunk 1]
```diff
@@ -12,25 +12,25 @@
 
 #include <Ice/Locator.ice>
 
-/// IceLocatorDiscovery is an Ice plug-in that enables the discovery of IceGrid and custom locators via
-/// UDP multicast.
-/// The IceDiscovery plug-in implements the {@see Ice::Locator} interface to locate (or discover) locators such as
-/// the IceGrid registry or custom IceGrid-like locator implementations using UDP multicast.
+/// LocatorDiscovery is an Ice plug-in that enables the discovery of IceGrid and custom locators via UDP multicast.
```
Comment (externl) [inline slice/IceLocatorDiscovery/IceLocatorDiscovery.ice:15]: We should add a TODO to remove this flie.

**Review by externl — APPROVED** [review]
Looks good to me

**Pull request merged**